### ISSUE TITLE
Dual-approach for the TOC

### ISSUE THREAD

**Issue by HadrienGardeur** [w3c/wpub#350]
As a follow-up to our discussions at TPAC, I think it's worth re-opening the case of the dual-approach for the TOC.

While this idea was proposed before (https://github.com/w3c/wpub/issues/291#issuecomment-416344052, https://github.com/w3c/wpub/issues/291#issuecomment-416352888, https://github.com/w3c/wpub/issues/291#issuecomment-416356123), it still seems very relevant following comments from Penguin Random House and Hachette Livre (respectively the largest trade publishers in the US and in France) that this is how they're producing EPUB files today.

If my understanding is correct (cc @laudrain and @liisamk) this is how they handle their EPUB production today:

- a navigation document (machine readable TOC) is produced but excluded from the reading order, it is only meant to be parsed and not displayed directly
- a more visually rich TOC is created separately, included in the reading order but it's not identifiable as such

Our current attempt in EPUB and in this group to use the same document for rendering and machine readable info still feels like a dead end IMO. No matter how we define an algorithm for extracting the TOC, we always reach a limit where things eventually fail.
Publishers might also decide that they want a radically different navigation (for example a map as an SVG) or add a lot of additional information that should not be processed (@liisamk mentioned such examples at TPAC).

This makes me wonder if we're not trying to use the term TOC for two separate concepts:

- a machine-readable document, mean to be processed by UAs to facilitate navigation
- a (potentially) visually-rich document, primarily meant to look good to the reader and that isn't necessarily text-based

While some publications may be able to use the same document for both purposes, I don't think that's true for all publications, far from it (as we can see with content produced by some of the largest trade publishers).

I've updated an earlier example to further illustrate how this could work:

- [Entry page](https://hadriengardeur.github.io/webpub-manifest/examples/why/)
- [Manifest](https://hadriengardeur.github.io/webpub-manifest/examples/why/wp.json)
- [TOC](https://hadriengardeur.github.io/webpub-manifest/examples/why/toc.html)

The idea is fairly basic:

- the machine readable TOC is identified by the `contents` rel value
- the "visually rich" navigation document is identified by the `navigation` rel value

The UAs could potentially do the following:

- provide the ability to jump to the "visually rich" navigation document
- parse and use the content of the machine readable document in its own affordances
- if the machine readable TOC is the only one present but can't be parsed, it would be rendered instead
- if both are presents and the machine readable document can't be parsed, the visually rich document would be rendered instead

**Comment by HadrienGardeur** [comment]
@iherman I haven't looked into it yet but it's not clear that something else than `doc-toc` would be needed.

**Comment by laudrain** [comment]
The answer is no, neither in [Digital Publishing roles](https://www.w3.org/TR/2017/REC-dpub-aria-1.0-20171214/) nor in [ARIA roles](https://www.w3.org/TR/wai-aria-1.1/).
But that is normal as role is for accessibility not for semantics.

**Comment by liisamk** [comment]
Hadrien-

You are correct that Hachette and PRH are both creating two docs with somewhat different purposes. This summary explains the goals/issues we have at PRH and I suspect is similar to what the experience is for Hachette or other Trade and Educational publishers.


  *   The navigation doc/machine readable TOC is meant to be as rich as possible and help a reader easily navigate through the text
     *   It often has more levels of navigation than would exist in print
        *   Examples: recipes, charts, subheads/sections within business books
     *   Current issues:
        *   There is not consistent support for 3 or more levels across reading systems
        *   We would love to have support for a minimal amount of styling- essentially Bold and Italic so that we can call out Titles of other works

  *   A more visually rich TOC that is often more of an introduction to the book and its content
     *   This may have more explanatory text that is not in the navigation
     *   It may be a higher level and not have as much detail as what is in the navigation
     *   It may be illustrated or even pictoral
     *   It may not exist in the digital product because it was not included in print
     *   It is typically identified in the guide with a type=”toc”
     *   It is included in the spine

I think it is rare that you would want to use the same doc for both of these purposes.

Let me know if it would help to have examples to see how this works.

Thanks!

Liisa

**Comment by HadrienGardeur** [comment]
Thanks @liisamk this is very helpful.

@llemeurfr @liisamk do you have examples that could be publicly shared with this group? This would be very helpful to further illustrate and understand this use case.

> [...] It is typically identified in the guide with a type=”toc”

There's no equivalent currently for the guide and its various types in WP. If some of them are useful (which seems to be the case here), we would need to either identify or define equivalent `rel` values.

**Comment by HadrienGardeur** [comment]
@iherman 

I'm not entirely convinced that we need an ARIA role for the visually-rich use case. Currently in EPUB, there's no `epub:type` for it.
It would probably be better in terms of accessibility, but we don't need it for the kind of additional semantics that we require from `doc-toc` (identifying the HTML element to extract its content).

Defining a `rel` value would be the equivalent of the `guide` property that currently exists in EPUB.

> Looking ahead, it is also not clear what the WebIDL representation should be.

IMO nothing is required.

I've already said this previously, but duplicating information is not a good thing and since we're not considering making this document machine processable, we don't need a place to store the info once processed.

**Comment by iherman** [comment]
> I'm not entirely convinced that we need an ARIA role for the visually-rich use case. Currently in EPUB, there's no epub:type for it.
It would probably be better in terms of accessibility, but we don't need it for the kind of additional semantics that we require from doc-toc (identifying the HTML element to extract its content).

I actually do not necessarily care about accessibility in this case, only a 'sign' that signals that this is a visually rich entry. Please look at my summary of the current process (see above): the role of `doc-toc` is not used as an accessibility hook, "just" as identification. This is what I am talking about.

> I've already said this previously, but duplicating information is not a good thing and since we're not considering making this document machine processable, we don't need a place to store the info once processed.

A visually navigation should be treated like ToC and and the page list. This means that, obviously, results of #338 are relevant for this entry as well. That issue is still open.

**Comment by HadrienGardeur** [comment]
> I actually do not necessarily care about accessibility in this case, only a 'sign' that signals that this is a visually rich entry.

The `rel` values already handles that: https://github.com/HadrienGardeur/webpub-manifest/blob/gh-pages/examples/why/wp.json#L27

**Comment by laudrain** [comment]
@HadrienGardeur we have an epub:type for the visual table of content. 
In Hachette, for EPUB3 today, we use epub:type="toc" in the spine document with the visual TOC.
This is not the nav document.

**Comment by HadrienGardeur** [comment]
@laudrain is there any valid reason not to use `doc-toc` then?

The two documents would still be properly identified by different `rel` values in the manifest.

**Comment by laudrain** [comment]
We map role="doc-toc" from epub:type="toc".

**Comment by iherman** [comment]
>>I actually do not necessarily care about accessibility in this case, only a 'sign' that signals that this is a visually rich entry.
>
> The rel values already handles that: https://github.com/HadrienGardeur/webpub-manifest/blob/gh-pages/examples/why/wp.json#L22

Please look at https://github.com/w3c/wpub/issues/350#issuecomment-432508128. The current draft does not allow fragmented URI-s for resources, so the rel value is not enough. Or we have to relax that restriction overall.

**Comment by iherman** [comment]
> @laudrain is there any valid reason not to use doc-toc then?

`doc-toc` has a specific meaning in ARIA. This may not be valid for the navigation document.

---

There were some discussions at TPAC and it _may_ (I said _MAY_) be possible that a new, general `purpose` attribute would be introduced in HTML that does not have the ARIA baggage to it. We could then define terms to be used for that new attribute, and this may solve the issue.

No commitment, though.

**Comment by laudrain** [comment]
> doc-toc has a specific meaning in ARIA. This may not be valid for the navigation document.

@iherman I'll check with my DAISY friends

**Comment by iherman** [comment]
@HadrianGardeur

Indeed, we seem to mutually not understanding one another, as it seems.

In my understanding, this is what you propose:

```json
"resources" : [{
        ...
    },{
        "type":"PublicationLink",
        "url": "http://www.example.org/nav.html",
        "rel": "navigation"
    }],
```

What happens if there are several nav elements in `nav.html`? What happens if you have a TOC _as well as_ visual rich navigation element in `http://www.example.org/nav.html`? How would you find the right navigation element?

Your proposal only works if `http://www.example.org/nav.html` is _only_ used for (a single) visually rich navigation. I do not think this restriction is acceptable. Alternatively, you have to use `http://www.example.org/nav.html#visuallyrichnavigation` in the Publication Link. This is against the current draft requirement.

**Comment by HadrienGardeur** [comment]
@iherman 

> What happens if there are several nav elements in `nav.html`?

Since that document is meant to be rendered, I think that's not really an issue. From an affordance perspective, I think rendering the document is enough, no need to jump to a specific fragment id.

> What happens if you have a TOC as well as visual rich navigation element in `http://www.example.org/nav.html`?

They're not semantically the same thing and IMO should be tied to different affordances as well.

[In my first post](https://github.com/w3c/wpub/issues/350#issue-372852253), I suggested the following behavior:
> - provide the ability to jump to the "visually rich" navigation document
> - parse and use the content of the machine readable document in its own affordances
> - if the machine readable TOC is the only one present but can't be parsed, it would be rendered instead
> - if both are presents and the machine readable document can't be parsed, the visually rich document would be rendered instead

I'll let @laudrain and @liisamk chime in to make sure that this is aligned with their expectations.

> How would you find the right navigation element?

I don't think that's necessary.

To summarize my proposal:

- in addition to the TOC and page list, I'd like to introduce the concept of resources meant for "navigation", a generic solution to cover all other navigation use cases
- unlike TOC and page list, there's no expectation that the UA would process such resources to extract any kind of structure from them
- the only affordance expected from the UA would be to jump at any time to such navigation resources and render them as is

I think this is aligned with a real-world use case, as it's been expressed during TPAC by major trade publishers.

**Comment by iherman** [comment]
Well... in my view, they way you put it, this is grossly underspecified in my view. The only think we specify is one resource that _MAY_ include several visually rich navigation elements without specifying which one should be used by the UA as part of its specific (and obviously different) affordance for navigation, that _MAY_ be present alongside the navigation meant for specific TOC which may lead to a possible confusion for the UA, etc.

You do not think it is necessary to specify it more precisely, I disagree. At this point we should agree that we disagree, and I let @laudrain, @llemeurfr, @liisamk, and the others decide whether such a loosely specified thing is useful or not.

**Comment by llemeurfr** [comment]
I'll try to put it differently: this issue seems to me similar to the cover-image vs cover-page issue. A printed book as a cover page, it has a ToC, both can be categorized as "visually rich". A WP-aware UA (i.e. a reading system) needs some machine processable data, as simple to process as possible: a cover image (optional) to be used into a catalog view or a book description; and a ToC (optional but recommended) to be displayed at a fingertip without leaving the current page. 

The cover page and visually rich TOC are both usually accessed in reading order. The UA has no real need to get a specific reference to them. 

Publishers tell us that this machine ToC is often different from the visually rich ToC (the former is often more complete than the latter). The UA needs a specific reference to the machine ToC and will render it with a UA specific layout (often in a panel). It will use the rel=contents solution and the corresponding structure will get an ARIA doc-toc attribute, because people using specific a11y tools need to access a complete ToC also. 

I conclude that using rel=contents is sufficient for publishers and reading system needs, and we don't need to define any rel=navigation for help accessing the visually rich ToC.

And this should bring peace to @iherman and @HadrienGardeur :-)

Same for a page list and other landmarks.

**Comment by HadrienGardeur** [comment]
@iherman I'm trying to be very pragmatic here and simply stick to the requirements of the industry.

If @laudrain and @liisamk both believe that:

- we don't need an affordance to jump directly to such a "visually-rich" navigation document
- there's no need for additional navigation aside from the machine-readable TOC and the page list

... then I'm happy to close this issue.

**Comment by avneeshsingh** [comment]
Regarding aria doc-toc attribute, it is for both semantics and accessibility (until we have a separate semantics related attribute available in HTML). 
So, it should be present in both machine readable TOCs as well as visually rich navigation document (if we go on this path) to enable AT recognize that the document provides navigation structure.
